llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -18738,8 +18738,9 @@ SDValue DAGCombiner::rebuildSetCC(SDValue N) {
       if (LegalTypes)
         SetCCVT = getSetCCResultType(SetCCVT);
       // Replace the uses of XOR with SETCC
-      return DAG.getSetCC(SDLoc(N), SetCCVT, Op0, Op1,
-                          Equal ? ISD::SETEQ : ISD::SETNE);
+      const ISD::CondCode CC = Equal ? ISD::SETEQ : ISD::SETNE;
+      if (!LegalOperations || TLI.isCondCodeLegal(CC, Op0.getSimpleValueType()))
+        return DAG.getSetCC(SDLoc(N), SetCCVT, Op0, Op1, CC);
```

Comment:
This could use an explanation. Previously we'd return SETCC unconditionally. Now we're only doing it... sometimes? While I know what `TLI.isCondCodeLegal` is for, it's not quite obvious how/why our decision here depends on `LegalOperations`.

Reply:
I'm disabling this particular transform from running after operation legalization on targets where the SETCC which would be created is not legal. I'm allowing the fold to happen prior to operation legalization in case this folding could trigger other transformations which might be beneficial.